test/unit/libraries/string-transform.spec.js

```
@@ -0,0 +1,83 @@
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+// @flow
+
+import { describe, it, expect, beforeEach } from '../../helpers/dependencies'
+import {
+  applyTransformation,
+  filterByString,
+  getCurrentTimeISOFormat,
+  prependWithFn
+} from '../../../src/libraries/string-transform'
+import { CallbackRecorder } from '../../helpers/utils'
+
+describe('prependWithFn', () => {
+  it('prepends each time with fn execution result', () => {
+    let calledTimes = 0
+    const prependFn = () => {
+      calledTimes += 1
+      return calledTimes.toString()
+    }
+    const transform = prependWithFn(prependFn)
+    expect(transform('data')).to.eql('1data')
+    expect(transform('data')).to.eql('2data')
+  })
+})
+
+describe('filterByString', () => {
+  it('returns same string if filter matches', () => {
+    expect(filterByString('a')('bald')).to.eql('bald')
+  })
+  it('returns nothing if filter doesn\'t match', () => {
+    expect(filterByString('a')('bold')).to.be.undefined
+  })
+})
+
+describe('applyTransformation', () => {
+  let cbRec
+
+  beforeEach(() => {
+    cbRec = new CallbackRecorder()
+  })
+
+  it('runs transformation function and pipes its output to callback', () => {
+    const transform = (data) => data + '+'
+
+    applyTransformation(transform, cbRec.getCallback())('data')
+    expect(cbRec.invoked).to.be.true
+    expect(cbRec.argument).to.eql('data+')
+  })
+
+  it('does not call callback if transformation returns undefined or null', () => {
```

> `null || undefined` is an expected as a return type from transformation, in that case we don't call the `next()` callback 

This hides errors from us - transformation unexpectedly and there are no warnings about that.

We should at least log unexpected behaviour. Or maybe even throw an exception?

---

it is an expected behavior, if an error is thrown inside transformation then it bubbles up

---

Yeah, it `tail` will get that error, not us. Can we have error callback or at least log this error?